### PR TITLE
Initial implementation of flag /cbstring

### PR DESCRIPTION
**Pull request by yasster**

### CODE REVIEW

File: clang/include/clang/Driver/Options.td [hunk 1]
```diff
@@ -9137,6 +9141,8 @@ def _SLASH_QIntel_jcc_erratum : CLFlag<"QIntel-jcc-erratum">,
   Alias<mbranches_within_32B_boundaries>;
 def _SLASH_arm64EC : CLFlag<"arm64EC">,
   HelpText<"Set build target to arm64ec">;
+def _SLASH_cbstring : CLFlag<"cbstring">, Alias<fms_cbstring>,
```
Comment (dpaoliello) [inline clang/include/clang/Driver/Options.td:9144]: Add a test to `clang/test/Driver/cl-options.c` for this.

File: clang/test/Driver/cbstring-simple.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    
+    // Use some strings to prevent optimization
+    printf("%s %s %s\n", global_str, array[0], main_str);
+    test_function();
+    
+    return 0;
+}
+
+// With /cbstring flag, strings should be in .text sections
+// CBSTRING: .text
```
Comment (dpaoliello) [inline clang/test/Driver/cbstring-simple.c]: If you change the `FileCheck` command line above to have `--check-prefixes=NORMAL,CHECK` and `--check-prefix=CBSTRING,CHECK` then you can do something like:

```
// NORMAL: .rdata
// CBSTRING: .text
// CHECK: global_test
// NORMAL: .rdata
// CBSTRING: .text
// CHECK: array_element1
```

You could even go a bit crazy and leverage the [substitution capture support](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks) to be able to use `CHECK` in most places:
```
// NORMAL: [[section|\.rdata]]
// CBSTRING: [[section|\.text]]
// CHECK: global_test
// CHECK: [[section]]
// CHECK: array_element1
```

File: clang/test/Driver/cbstring-simple.c [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+// CBSTRING: global_test
```
Comment (dpaoliello) [inline clang/test/Driver/cbstring-simple.c]: If the text is immediately after the section declaration in the output, then please use `CBSTRING-NEXT`

File: clang/test/Driver/cbstring-simple.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
+// Comprehensive test for /cbstring flag functionality with all edge cases
+// RUN: %clang_cl /cbstring /c %s -o %t_cbstring.obj
```
Comment (dpaoliello) [inline clang/test/Driver/cbstring-simple.c]: Be careful using a slash at the front of command line arguments, it can cause issues on other platforms. I recommend switching to using the regular `clang` driver and dashes.

File: clang/test/Driver/cbstring-simple.c [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+// RUN: %clang_cl /c %s -o %t_normal.obj  
+// RUN: llvm-objdump -t %t_cbstring.obj | FileCheck %s --check-prefix=CBSTRING
```
Comment (dpaoliello) [inline clang/test/Driver/cbstring-simple.c]: Can we use `-emit-llvm` and check the IR instead of building and dumping an objdump? (Not sure objdump is guaranteed to be built as part of the `check-clang` target).

File: clang/test/Driver/cbstring-test.c [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// Test file for /cbstring flag functionality
+// RUN: %clang_cl /cbstring /c %s -o %t_cbstring.obj
+// RUN: %clang_cl /c %s -o %t_normal.obj  
+// RUN: llvm-objdump -t %t_cbstring.obj | FileCheck %s --check-prefix=CBSTRING
+// RUN: llvm-objdump -t %t_normal.obj | FileCheck %s --check-prefix=NORMAL
+
+// Test that string literals are placed in .text section with /cbstring flag
```
Comment (dpaoliello) [inline clang/test/Driver/cbstring-test.c]: How is this different to the other test?

File: clang/test/Driver/cbstring-simple.c [hunk 7]
```diff
@@ -0,0 +1,79 @@
+// Comprehensive test for /cbstring flag functionality with all edge cases
```
Comment (dpaoliello) [inline clang/test/Driver/cbstring-simple.c]: I don't think that the Driver directory is the correct place for this kind of test, CodeGen seems like it would be a better fit.

**Review by dpaoliello — CHANGES_REQUESTED** [review]
Implementation is good, tests need cleanup

File: clang/test/Driver/cl-options.c [hunk 8]
```diff
@@ -198,6 +198,14 @@
...
 // Ox: -O3
+
+// RUN: %clang_cl /cbstring --target=i686-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring %s
+// RUN: %clang_cl /cbstring --target=x86_64-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring %s
+// cbstring: -fms-cbstring
+
+// RUN: %clang_cl --target=i686-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring-default %s
+// RUN: %clang_cl --target=x86_64-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring-default %s
+// cbstring-default-NOT: -fms-cbstring
 // Ox: -mframe-pointer=none
```
Comment (dpaoliello) [inline clang/test/Driver/cl-options.c]: Bad merge? (Should this be above your change?)

File: clang/test/Driver/cl-options.c [hunk 9]
```diff
@@ -198,6 +198,14 @@
...
 // RUN: %clang_cl /Ox --target=x86_64-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=Ox %s
 // Ox: -O3
+
+// RUN: %clang_cl /cbstring --target=i686-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring %s
+// RUN: %clang_cl /cbstring --target=x86_64-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring %s
+// cbstring: -fms-cbstring
+
+// RUN: %clang_cl --target=i686-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring-default %s
+// RUN: %clang_cl --target=x86_64-pc-windows-msvc -### -- %s 2>&1 | FileCheck -check-prefix=cbstring-default %s
+// cbstring-default-NOT: -fms-cbstring
```
Comment (dpaoliello) [inline clang/test/Driver/cl-options.c]: This file doesn't usually have negative tests, except checking which flags are enabled at different optimization levels.

File: clang/test/CodeGen/ms-cbstring.c [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+// Array elements  
+// NORMAL: constant {{.*}} c"array_element1\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"array_element1\00", section ".text$s"
+// NORMAL: constant {{.*}} c"array_element2\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"array_element2\00", section ".text$s"
+
+// Strings in regular functions go to .text$s
+// NORMAL: constant {{.*}} c"local_test\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"local_test\00", section ".text$s"
```
Comment (dpaoliello) [inline clang/test/CodeGen/ms-cbstring.c]: Are the strings no longer comdats?

File: clang/test/CodeGen/ms-cbstring.c [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    printf("%s\n", static_main_str);
+    test_function();
+    in_other_seg();
+    
+    return 0;
+}
+
+// Check string constants in order they appear in IR:
+
+// Global strings
```
Comment (dpaoliello) [inline clang/test/CodeGen/ms-cbstring.c]: Can you please verify if MSVC places global strings into `.text$s`? From my own testing it didn't...

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 12]
```diff
@@ -7005,7 +7014,7 @@ ConstantAddress CodeGenModule::GetAddrOfConstantCString(const std::string &Str,
 
   // Create a global variable for this.
   auto GV = GenerateStringLiteral(C, llvm::GlobalValue::PrivateLinkage, *this,
-                                  GlobalName, Alignment);
+                                  GlobalName, Alignment, nullptr);
```
Comment (dpaoliello) [inline clang/lib/CodeGen/CodeGenModule.cpp]: Leverage the default arg?

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 13]
```diff
@@ -6908,9 +6908,17 @@ GenerateStringLiteral(llvm::Constant *C, llvm::GlobalValue::LinkageTypes LT,
...
   CGM.setDSOLocal(GV);
 
-  // If /cbstring flag is enabled, place string literals in .text section
+  // If /cbstring flag is enabled, place string literals in function-specific sections
+  // following MSVC behavior: append $s to current function's section, or use .text$s
   if (CGM.getCodeGenOpts().CBString) {
-    GV->setSection(".text");
+    if (CurrentFunc && CurrentFunc->hasSection()) {
+      // If function has a section, append $s to it
+      std::string FuncSection = CurrentFunc->getSection().str();
```
Comment (dpaoliello) [inline clang/lib/CodeGen/CodeGenModule.cpp]: Not sure if this is the preferred pattern, or making a `Twine` and then converting that into a `StringRef` (which I believe uses a `std::string` under the covers anyway). Just a heads up in case someone makes a note in the public review.

**Review by dpaoliello — APPROVED** [review]
Let's make sure we are matching MSVC's behavior. Otherwise looks good.

**Comment by dpaoliello** [comment]
One other note, you may want to consider using the `-SAME` check in FileCheck to reduce the duplication in the tests: https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-same-directive

File: clang/test/Driver/cl-options.c (hunk 8 above)
Comment (yasster) [reply clang/test/Driver/cl-options.c]: Yeah just a bad merge, I'll fix it

File: clang/test/CodeGen/ms-cbstring.c (hunk 10 above)
Comment (yasster) [reply clang/test/CodeGen/ms-cbstring.c]: They are, I'll include it in the test

File: clang/test/CodeGen/ms-cbstring.c [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// Check string constants in order they appear in IR:
+// Note: With -fms-cbstring, function strings get section attributes but remain comdats for proper deduplication
+// Note: Global strings keep default section behavior (like MSVC's CONST segment)
+
+// Global strings - should keep default section behavior (not go to .text$s)
+// NORMAL: constant {{.*}} c"global_test\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"global_test\00"{{.*}}comdat{{.*}}align
+// CBSTRING-NOT: constant {{.*}} c"global_test\00"{{.*}}section
```
Comment (dpaoliello) [inline clang/test/CodeGen/ms-cbstring.c]:
```suggestion
// CHECK: constant {{.*}} c"global_test\00"
// CHECK-NOT: section
// CHECK-SAME: comdat
```

Otherwise `-NOT` checks that the lines doesn't appear AFTER the last `CHECK`, so this isn't a useful check. Also, the `{{.*}}` will eat any section declaration, so this will still match if there is an explicit section.

File: clang/test/CodeGen/ms-cbstring.c [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+// NORMAL: constant {{.*}} c"array_element1\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"array_element1\00"{{.*}}comdat{{.*}}align
+// CBSTRING-NOT: constant {{.*}} c"array_element1\00"{{.*}}section
+// NORMAL: constant {{.*}} c"array_element2\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"array_element2\00"{{.*}}comdat{{.*}}align
+// CBSTRING-NOT: constant {{.*}} c"array_element2\00"{{.*}}section
+
+// Function local strings go to .text$s
+// NORMAL: constant {{.*}} c"local_test\00"{{.*}}comdat
+// CBSTRING: constant {{.*}} c"local_test\00", section ".text$s"{{.*}}comdat
```
Comment (dpaoliello) [inline clang/test/CodeGen/ms-cbstring.c]:
```suggestion
// CHECK: constant {{.*}} c"local_test\00"
// CBSTRING-SAME: section ".text$s"
// NORMAL-NOT: section
// CHECK-SAME: comdat
```

Same with this, the `{{.*}}` for `NORMAL` will match against `, section ....` so it isn't a useful test.

**Review by dpaoliello — CHANGES_REQUESTED** [review]

**Review by dpaoliello — APPROVED** [review]
